source/alloc/src/heap.rs

```diff
@@ -1,505 +1,128 @@
-            for i in 0..count {
-                start.add(i).write((f)());
-            }
-        }
-
-        Ok(HeapArray {
-            ptr: aa_ptr,
-            pd: PhantomData,
-        })
+    pub unsafe fn init(&self, start: NonNull<u8>, len: usize) {
+        self.allocator.init(start, len)
     }
+}
 
-    pub fn alloc_fixed_vec<T>(&mut self, capacity: usize) -> Result<HeapFixedVec<T>, ()> {
-        // Clean up any pending allocs
-        self.clean_allocs();
-
-        // Then figure out the layout of the requested array. This call fails
-        // if the total size exceeds ISIZE_MAX, which is exceedingly unlikely
-        // (unless the caller calculated something wrong)
-        let layout = unsafe { ActiveArr::<MaybeUninit<T>>::layout_for_arr(capacity) };
-
-        // Then, attempt to allocate the requested T.
-        let nnu8 = self.get_heap().allocate_first_fit(layout)?;
-        let aa_ptr = nnu8.cast::<ActiveArr<MaybeUninit<T>>>();
-
-        // And initialize it with the contents given to us
-        unsafe {
-            ActiveArr::<MaybeUninit<T>>::write_heap(aa_ptr, self.aheap);
-            ActiveArr::<MaybeUninit<T>>::write_capacity(aa_ptr, capacity);
-            let (start, count) = ActiveArr::<MaybeUninit<T>>::data(aa_ptr);
-            let start = start.as_ptr();
-            for i in 0..count {
-                start.add(i).write(MaybeUninit::uninit());
-            }
+unsafe impl<U: UlAlloc> GlobalAlloc for AHeap2<U> {
+    #[inline(always)]
+    unsafe fn alloc(&self, layout: core::alloc::Layout) -> *mut u8 {
+        if INHIBIT_ALLOC.load(Ordering::SeqCst) {
+            // TODO
+            return null_mut();
         }
-
-        Ok(HeapFixedVec {
-            ptr: aa_ptr,
-            pd: PhantomData,
-            len: 0,
-        })
-    }
-
-    pub fn alloc_raw(&mut self, layout: Layout) -> Result<NonNull<()>, ()> {
-        // Clean up any pending allocs
-        self.clean_allocs();
-
-        // calculate the layout of the requested allocation
-        let (layout, offset) = ActiveUnsized::layout(layout);
-
-        // Then, attempt to allocate the requested T.
-        let nnu8 = self.get_heap().allocate_first_fit(layout)?;
-        let ptr = nnu8.cast::<ActiveUnsized>();
-
-        unsafe {
-            ActiveUnsized::write_heap(ptr, self.aheap);
-            let data_ptr = ptr.as_ptr().cast::<u8>().add(offset).cast::<()>();
-            Ok(NonNull::new_unchecked(data_ptr))
+        let ptr = self.allocator.alloc(layout);
+        if ptr.is_null() {
+            INHIBIT_ALLOC.store(true, Ordering::SeqCst); // TODO
         }
+        ptr
     }
-}
 
-impl Drop for HeapGuard {
-    fn drop(&mut self) {
-        self.aheap.state.store(AHeap::INIT_IDLE, Ordering::SeqCst)
+    unsafe fn dealloc(&self, ptr: *mut u8, layout: core::alloc::Layout) {
+        self.allocator.dealloc(ptr, layout);
+        INHIBIT_ALLOC.store(false, Ordering::SeqCst); // TODO
+        OOM_WAITER.wake_all();
```

Comment:
a thought for the future: eventually, we probably shouldn't wake *every* task that's waiting on an allocation in `dealloc`, since the freed memory will probably not be sufficient to satisfy every outstanding allocation request. Instead, it would be nice to have the waiting node for each allocation request store the layout it desires, and then traverse the linked list until we find a waiter that can be satisfied by the freed layout, subtract its desired layout from the freed block, and (if there's still space in the layout) continue waking tasks.

but, this is substantially more complex, since it can't easily be modeled with `maitake`'s existing synchronization primitives (although, it could be a potential extension to `WaitMap` maybe?). for now, this is fine, but it's maybe worth revisiting.